### PR TITLE
[DRAFT] Conditional discrepancy tests - Kernel approach and Bregman divergence

### PR DESCRIPTION
**Pull request by adam2392**
First step for #56 

Changes proposed in this pull request:

Implements two new conditional discrepancy (CD) tests
- Kernel CD test: basically KCI but for testing `p_1(y|x) = p_2(y|x)`
- Bregman divergence CD test: uses divergences to test `p_1(y|x) = p_2(y|x)`

Still need to add unit tests. Once I get CI and tests working, I'll split this up into two PRs.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: dodiscover/cd/bregman.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    ):
+        rng = np.random.default_rng(random_state)
+
+        p = 0.5
+        n_samps = X.shape[0]
+
+        # compute the test statistic on the conditionally permuted
+        # dataset, where each group label is resampled for each sample
+        # according to its propensity score
+        null_dist = Parallel(n_jobs=self.n_jobs)(
```
Comment (darthtrevino) [inline dodiscover/cd/bregman.py:136]: What parallelism mechanisms do we have currently in the library? Can users configure/tune this in any way?

File: dodiscover/cd/kernel_test.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        # TODO: CHECK THAT THIS IS CORRECT
+        K0 = K[np.ix_(first_mask, first_mask)]
+        K1 = K[np.ix_(second_mask, second_mask)]
+
+        # compute the number of samples in each
+        n0 = int(np.sum(1 - z))
+        n1 = int(np.sum(z))
+
+        if isinstance(self.l2, (int, float)):
```
Comment (darthtrevino) [inline dodiscover/cd/kernel_test.py:225]: It kind of seems like this logic should be packed in a property getter. `self.get_regs()`

File: dodiscover/ci/kernel_test.py [hunk 3]
```diff
@@ -151,24 +151,31 @@ def test(
             # when running CI, \ddot{X} comprises of (X, Z)
             X = np.concatenate((X, Z), axis=1)
 
-            Kz, sigma_z = self._compute_kernel(
+            Kz, sigma_z = compute_kernel(
                 Z,
-                distance_metric="l2",
+                distance_metric="euclidean",
```
Comment (darthtrevino) [inline dodiscover/ci/kernel_test.py:156]: Would an enum make more sense than a string?

**Comment by adam2392** [comment]
> Some minor nits, I think we should look into using a set of domain enumerations instead of depending on hardcoded string values for metrics and distance metrics.

Btw, migrating this PR into two separate PRs: #81 and #82

File: dodiscover/cd/kernel_test.py (hunk 2 above)
Comment (adam2392) [reply dodiscover/cd/kernel_test.py:225]: Will modify in #81

File: dodiscover/cd/bregman.py (hunk 1 above)
Comment (adam2392) [reply dodiscover/cd/bregman.py:136]: Parallelism is mainly done via `joblib`, so we do not use anything beyond the basic functionality. 

There are definitely many things we can add, such as local caching (to allow a user to warm-start their causal discovery algo.), or mem-mapping if they have a gigantic dataset, but these are relatively low priority imo until we get the algos/API implemented(?).

I might be missing other functionalities modern joblib has

**Pull request closed without merging**